feat(gateway): CEL conditions for plugins (#5449)
Kong/developer.konghq.com: app/gateway/plugins/expressions.md

---
title: Conditional expressions for plugins

description: Reference for the CEL expression language used in {{site.base_gateway}} plugin conditions.
content_type: reference
layout: reference

products:
  - gateway

min_version:
  gateway: '3.15'

breadcrumbs:
  - /gateway/
  - /gateway/entities/
  - /gateway/entities/plugin/

faqs:
  - q: Do conditions work with global plugins?
    a: Yes, conditions can be used on global plugins that are not scoped to a Route, Service, Consumer, or Consumer Group.
  - q: When should I use plugin conditions instead of Routes?
    a: |
      Routes with expression router conditions should be used instead of per-plugin conditions wherever practical,

>       since Route expressions are more performant than plugin conditions.

Reviewer: ```suggestion
      Routes with expression router conditions should be used instead of per-plugin conditions, since Route expressions are more performant than plugin conditions.
```

I think "wherever practical" should be defined by us or not included ( I know this was here before this PR).

      This is because:
       * The `init` phase of plugins on excluded Routes won't execute.
       * The plugin condition won't need to be evaluated.

>   - q: Can I match a plugin condition based on request content type (for example, JSON or XML)?
>     a: |
>       The condition expression language doesn't support this explicitly.
>       You could use a plugin such as Datakit or Pre-Function to parse the body, extract the required value, and store it in `kong.ctx.shared`.
>       The plugin condition can then reference that `kong.ctx.shared` key.

Reviewer: ```suggestion
  - q: Can I match a plugin condition based on request content type (for example, JSON or XML)?
    a: |
      No, the condition expression language doesn't support this explicitly.
     As an alternative, you can use a plugin such as [Datakit](/plugins/datakit/) or [Pre-Function](/plugins/pre-function/) to parse the body, extract the required value, and store it in [`kong.ctx.shared`](/gateway/plugins/expressions/#kong-ctx-shared-fields).
      The plugin condition can then reference that `kong.ctx.shared` key.
```

  - q: What happens if my condition expression has a runtime error?
    a: |

>       If a condition expression fails at runtime, {{site.base_gateway}} logs the error at ERROR level and returns a 500 status code to the client.

Reviewer: ```suggestion
      If a condition expression fails at runtime, {{site.base_gateway}} logs the error at the ERROR level and returns a 500 status code to the client.
```

      To prevent this, wrap your expression in a `default()` call: `default(<expression>, false)`.
      This returns `false` (and skips the plugin) instead of a 500 if the expression errors.
  - q: |
      I was using ATC for plugin conditions in {{site.base_gateway}} 3.14. What do I need to change for 3.15?
    a: |
      The expression language changed between 3.14 and 3.15.
      In {{site.base_gateway}} 3.14, the feature was in beta and used ATC (Abstract Tree Classifier) syntax for plugin conditions. 
      For the 3.14 reference, see [Conditional expressions for plugins in 3.14](/gateway/plugins/expressions-314/).
      {{site.base_gateway}} 3.15 uses CEL (Common Expression Language), which isn't backwards-compatible.

>       Any condition expression that worked in 3.14 will need to be rewritten for 3.15.

Reviewer: ```suggestion
      Any conditional expression that worked in 3.14 will need to be rewritten for 3.15.
```

>       The main syntax changes are:

Reviewer: Is this worth an H2 in the body of this doc?

Author: I'm okay with moving it to that.

  - field: "`http.headers_list.<header_name>`"
    type: "`list<string>`"
    description: "All values of the specified request header as a list. Returns `null` if the header is absent."
    example: |
      `http.headers_list.x_roles != null && "editor" in http.headers_list.x_roles`
  - field: "`http.queries_list.<param_name>`"
    type: "`list<string>`"
    description: "All values of the specified query parameter as a list. Returns `null` if the parameter is absent."
    example: |
      `http.queries_list.tag != null && "featured" in http.queries_list.tag`
  - field: "`net.protocol`"
    type: "`string`"
    description: |
      The protocol of the Route, for example `"http"` or `"https"`.
    example: |
      `net.protocol == "https"`
  - field: "`net.tls.sni`"
    type: "`string`"
    description: "The server name from the TLS ClientHello packet, if the connection is over TLS. Returns `null` for non-TLS connections."
    example: |
      `net.tls.sni == "api.example.com"`
  - field: "`net.src.ip`"
    type: "`string`"
    description: "The IP address of the client."
    example: |
      `net.src.ip == "10.0.0.1"`
  - field: "`net.src.port`"

>     type: "`Int`"

Author: ```suggestion
    type: "`int`"
```

    description: "The port used by the client to connect."
    example: |
      `net.src.port > 1024`
  - field: "`net.dst.ip`"
    type: "`string`"
    description: "The listening IP address where {{site.base_gateway}} accepted the connection."
    example: |
      `net.dst.ip == "192.168.1.1"`
  - field: "`net.dst.port`"

>     type: "`Int`"

Author: ```suggestion
    type: "`int`"
```

  - title: Description
    key: description
  - title: Example
    key: example
rows:
  - field: "`plugins.<plugin_name>.is_matched`"
    type: "`boolean`"
    description: |
      * `true` if the plugin is configured and its Route or Service scope matches this request (Consumer scope excluded).
      * `false` if the plugin is configured but its scope does not match.
      * `null` if the plugin isn't configured.
    example: |
      `plugins.key_auth.is_matched == true`
  - field: "`plugins.<plugin_name>.priority`"
    type: "`int`"
    description: "The priority of the matched plugin. `null` if the plugin isn't matched."
    example: |
      `plugins.key_auth.priority > 1000`
  - field: "`plugins.<plugin_name>.access_has_executed`"
    type: "`boolean`"
    description: |
      * `true` if the plugin's `access` phase has already executed at the time this condition is evaluated.
      * `false` if the plugin is matched but its `access` phase has not run yet.
      * `null` if the plugin isn't configured.
    example: |
      `plugins.key_auth.access_has_executed == true`
{% endtable %}
<!--vale on-->

### Null handling

How null values behave depends on the field type.

> The following fields return `null` when the value isn't set:

Reviewer: ```suggestion
The following fields return `null` when a value isn't set:
```

* `http.headers.*`
* `http.queries.*`
* `http.headers_list.*`
* `http.queries_list.*`
* `net.tls.sni`
* All context fields (`consumer.*`, `route.*`, `service.*`, `principal.*`, `consumer_group.*`, `plugins.*`)

> You can compare these directly with `null`:

Reviewer: ```suggestion
You can compare these directly using `null`:
```

```

Alternatively, wrap the entire expression in `default()` to return a safe fallback if the key is missing:

```sh
default(kong.ctx.shared.my_flag == "enabled", false)
```

### Operators and functions

The following operators and functions are supported in plugin condition expressions:

<!--vale off-->
{% table %}
columns:
  - title: Operator or function
    key: type
  - title: Description
    key: description
rows:
  - type: "`&&`, `||`, `!`"
    description: "Logical AND, OR, NOT."
  - type: "`()`"
    description: "Group expressions to control evaluation order."
  - type: "`==`, `!=`, `<`, `<=`, `>`, `>=`"
    description: "Standard value comparison."
  - type: "`+`"
    description: "String concatenation."
  - type: "`in`"
    description: "Tests whether a value is a member of a list."
  - type: "`contains()`"
    description: "Returns `true` if the string contains the given substring."
  - type: "`starts_with()`"
    description: "Returns `true` if the string starts with the given prefix."
  - type: "`ends_with()`"
    description: "Returns `true` if the string ends with the given suffix."
  - type: "`matches()`"
    description: |

>       Tests the string against a [RE2 regular expression](https://docs.rs/regex/latest/regex/#syntax).

Author: ```suggestion
      Tests the string against a [RE2 regular expression](https://github.com/google/re2/wiki/Syntax).
```

      Matches any substring unless anchored with `^` and `$`.
  - type: "`size()`"
    description: "Returns the number of elements in a list, or the number of characters in a string."
  - type: "`has()`"
    description: "Returns `true` if the key exists in the map. Required before accessing inner keys of `kong.ctx.shared` or `principal.metadata`."
  - type: "`all()`"

>     description: "Returns `true` if all elements in the list satisfy the predicate."

Reviewer 2: Just a reminder: we haven't included examples for these comprehension operations (like `all()`, ... ). However, users can refer to the CEL specification documentation to learn how to write them.

Author: I'm going to leave this as-is without examples, exactly for that reason. We're already documenting a lot of CEL-specific content; examples for our own fields make sense to me, but this feels like it would be overkill.

Reviewer 2: Got it, make sense to leave it as-is for your explanation. So I resolved this.

  - type: "`exists()`, `map()`, `filter()`"
    description: "Additional CEL comprehension macros for working with lists."
  - type: "`default()`"
    description: |
      Wraps an expression to return a fallback boolean value if the expression produces a runtime error.

>       Must wrap the **entire** expression — cannot be used inline within a larger expression.

Reviewer 2: While this document provides various examples and explanations for `default()`, it is worth noting that its second argument only supports boolean values (true or false). Passing any other value will result in an error.

>       Must wrap the **entire** expression — cannot be used inline within a larger expression.

Author: ```suggestion
      Wraps an expression to return a fallback boolean value if the expression produces a runtime error.
      Must wrap the **entire** expression, and cannot be used inline within a larger expression.
      <br><br>
      The second argument for this function accepts only boolean values (`true` or `false`).
```

{% endtable %}
<!--vale on-->

{:.info}
> The `matches()` function uses [RE2 regex syntax](https://github.com/google/re2) and matches any substring by default.
> Use `^` to anchor the start and `$` to anchor the end of a pattern.

> > For example, `http.path.matches("^/api/v[0-9]+$")` matches `/api/v1` but not `/other/api/v1`.

Reviewer 2: since we have updated the ERD, see

<img width="729" height="166" alt="Image" src="https://github.com/user-attachments/assets/088dc60e-062e-452b-bf70-4acad9de2bec" />

The key point is: it is rust regular expression( similar to google re2), you could updated it with:

Reviewer 2: ```
Regular expressions follow the rules of [Rust Crate regex](https://docs.rs/regex/latest/regex/#syntax). Regular expression matches succeed if they match a substring of the argument. Use explicit anchors (^ and $) in the pattern to force full-string matching, if desired.
```

> > For example, `http.path.matches("^/api/v[0-9]+$")` matches `/api/v1` but not `/other/api/v1`.

Author: ```suggestion
> Regular expressions follow the rules of [Rust Crate regex](https://docs.rs/regex/latest/regex/#syntax). 
> Regular expression matches succeed if they match a substring of the argument. 
> Use explicit anchors (`^` and `$`) in the pattern to force full-string matching, if desired. 
> For example, `http.path.matches("^/api/v[0-9]+$")` matches `/api/v1` but not `/other/api/v1`.
```

### Types

Plugins support the following CEL types:

<!--vale off-->
{% table %}
columns:
  - title: Type
    key: type
  - title: Description
    key: description
  - title: Example literal
    key: example
rows:
  - type: "`bool`"
    description: "Boolean value."
    example: "`true`, `false`"
  - type: "`int`"
    description: "64-bit signed integer."
    example: "`42`, `-1`"
  - type: "`string`"
    description: "UTF-8 string."
    example: '`"hello"`'
  - type: "`list<string>`"
    description: "Ordered list of string values."
    example: '`["foo", "bar"]`'
  - type: "`map<string, _type>`"
    description: "Map with string keys and values of any type. Used for `kong.ctx.shared` and `principal.metadata`."
    example: "See `kong.ctx.shared` field."
  - type: "`null`"
    description: "Null value, returned when an optional field has no value."
    example: "`null`"
{% endtable %}
<!--vale on-->

### Handling default values

> Use `default()` to make a condition expression safe at runtime.

Reviewer: ```suggestion
`default()` makes condition expressions safe at runtime.
```
